### PR TITLE
Some perf

### PR DESCRIPTION
**Pull request by jasonLaster**
1. bail early for hasAwait
2. don't do unnecessary setModes
3. memoize Tab
4. memoize Frame

### CODE REVIEW

File: src/actions/pause/commands.js [hunk 1]
```diff
@@ -151,6 +151,19 @@ export function reverseStepOut() {
   };
 }
 
+/*
+ * Checks for await or yield calls on the paused line
+ * This avoids potentially expensive parser calls when we are likely
+ * not at an async expression.
+ */
+function hasAwait(source, pauseLocation) {
```
Comment (wldcordeiro) [inline src/actions/pause/commands.js]: This would be a good util to test.

File: src/utils/editor/source-documents.js [hunk 2]
```diff
@@ -126,7 +126,12 @@ function showSourceText(
   if (hasDocument(source.id)) {
     const doc = getDocument(source.id);
     if (editor.codeMirror.doc === doc) {
-      editor.setMode(getMode(source, symbols));
+      const mode = getMode(source, symbols);
+
+      if (doc.mode.name !== mode.name) {
```
Comment (wldcordeiro) [inline src/utils/editor/source-documents.js]: Nice idea to check probably saves a good amount of work too!

File: src/utils/source.js [hunk 3]
```diff
@@ -229,15 +229,18 @@ export function getSourceLineCount(source: Source) {
  * @static
  */
 
-export function getMode(source: Source, symbols?: SymbolDeclarations) {
+export function getMode(
+  source: Source,
+  symbols?: SymbolDeclarations
+): { name: string } {
```
Comment (wldcordeiro) [inline src/utils/source.js]: Nice consistency fix.

File: src/utils/editor/source-documents.js (hunk 2 above)
Comment (jasonLaster) [reply src/utils/editor/source-documents.js]: yeah, a ton of time

File: src/actions/pause/commands.js (hunk 1 above)
Comment (jasonLaster) [reply src/actions/pause/commands.js]: good call

**Comment by wldcordeiro** [comment]
Circle is failing due to linting/bad imports

```
yarn lint-js

yarn run v1.3.2
$ eslint *.js "src/**/*.js" --fix

/home/ubuntu/debugger.html/src/components/SecondaryPanes/Frames/index.js
  26:3  error  getCallStackFrames not found via ../../../selectors.js -> ../../../selectors/getCallStackFrames.js  import/named

/home/ubuntu/debugger.html/src/selectors.js
  30:10  error  getCallStackFrames not found in './selectors/getCallStackFrames'  import/named```

and Travis is failing on tests seemingly because of the missing selector function.

```
  TEST-UNEXPECTED-FAIL | devtools/client/debugger/new/test/mochitest/browser_dbg-asm.js | A promise chain failed to handle a rejection: _selectors.getCallStackFrames is not a function - stack: thunk/</</< devtools/client/debugger/new/debugger.js 44077:45
     bindActionCreator/< devtools/client/debugger/new/debugger.js 35486:12m
     paused devtools/client/debugger/new/debugger.js 43348:5
     async*eventSource/proto.emit devtools/shared/client/event-source.js 124:9
     _onThreadState devtools/shared/client/thread-client.js 619:35
     onPacket devtools/shared/client/debugger-client.js 911:7
     send/< devtools/shared/transport/transport.js 570:13
     exports.makeInfallible/< devtools/shared/ThreadSafeDevToolsUtils.js 109:14
     DevToolsUtils.executeSoon*exports.executeSoon devtools/shared/DevToolsUtils.js 57:19
     send devtools/shared/transport/transport.js 563:9
     send devtools/server/main.js 1496:5
     receiveMessage devtools/shared/transport/transport.js 761:7
   chrome://mochikit/content/browser-test.js:Tester.prototype.nextTest<:703
   timeoutFn@chrome://mochikit/content/browser-test.js:1160:9
   setTimeout handler*SimpleTest_setTimeoutShim@chrome://mochikit/content/tests/SimpleTest/SimpleTest.js:676:12
   timeoutFn@chrome://mochikit/content/browser-test.js:1153:15
   setTimeout handler*Tester_execTest@chrome://mochikit/content/browser-test.js:1122:9
   Tester.prototype.nextTest</<@chrome://mochikit/content/browser-test.js:956:9
   SimpleTest.waitForFocus/waitForFocusInner/focusedOrLoaded/<@chrome://mochikit/content/tests/SimpleTest/SimpleTest.js:795:59```

**Pull request merged**